api/src/services/messaging.js

```diff
@@ -0,0 +1,306 @@
+  return messages.filter(message => {
+    return requiredFields.every(field => {
+      if (!message[field]) {
+        logger.info(`Message missing required field "${field}": ${JSON.stringify(message)}`);
+      } else {
+        return true;
+      }
+    });
+  });
+};
+
+const removeDuplicateMessages = messages => {
+  if (!messages.length) {
+    return Promise.resolve([]);
+  }
+  const keys = messages.map(message => message.id);
+  return db.medic.query('medic-sms/messages_by_gateway_ref', { keys })
+    .then(res => res.rows.map(row => row.key))
+    .then(seenIds => messages.filter(message => {
+      if (seenIds.includes(message.id)) {
+        logger.info(`Ignoring message (ID already seen): "${message.id}"`);
+      } else {
+        return true;
+      }
+    }));
+};
+
+const validateIncomingMessages = messages => {
+  return removeDuplicateMessages(validateRequiredFields(messages));
+};
+
+const createDocs = messages => {
+  if (!messages.length) {
+    return [];
+  }
+  const docs = messages.map(message => records.createByForm({
+    from: message.from,
+    message: message.content,
+    gateway_ref: message.id,
+  }));
+  return config.getTransitionsLib().processDocs(docs);
+};
+
+const resolveMissingUuids = changes => {
+  const gatewayRefs = changes
+    .map(change => !change.messageId && change.gatewayRef)
+    .filter(ref => ref);
+  if (!gatewayRefs.length) {
+    // all messages have ids
+    return Promise.resolve();
+  }
+  return db.medic.query('medic-sms/messages_by_gateway_ref', { keys: gatewayRefs }).then(res => {
+    res.rows.forEach(({ key, value }) => {
+      const change = changes.find(({ gatewayRef }) => gatewayRef === key);
+      if (change) {
+        change.messageId = value;
+      }
+    });
+  });
+};
+
+module.exports = {
+  
+  /**
+   * Sends pending messages on the doc with the given ID using the
+   * configured outgoing message service.
+   */
+  send: docId => {
+    const service = getOutgoingMessageService();
+    if (!service) {
+      return Promise.resolve();
+    }
+    return db.medic.get(docId).then(doc => {
+      return sendMessages(service, getPendingMessages(doc));
+    });
+  },
+
+  /**
+   * Stores incoming messages in the database.
+   * @param messages an array of message objects with properties
+   *    - id: unique gateway reference used to prevent double handling
+   *    - from: the phone number of the original sender
+   *    - content: the string message
+   */
+  processIncomingMessages: (messages=[]) => {
+    return validateIncomingMessages(messages)
+      .then(messages => createDocs(messages))
+      .then(results => {
+        const allOk = results.every(result => result.ok);
+        if (!allOk) {
+          logger.error('Failed saving all the new docs: %o', results);
+          throw new Error('Failed saving all the new docs');
+        }
+      });
+  },
+
+  /*
+   * Returns `options.limit` messages, optionally filtering by state.
```

Comment:
This comment seem outdated by the refactoring.

Reply:
Fixed.